plasmapy/plasma/grids.py

```diff
@@ -601,18 +629,100 @@ def _make_mesh(self, start, stop, num, **kwargs):
+        grid points in all axes.
+
+        Parameters
+        ----------
+        pos : np.ndarray or u.Quantity array, shape (n,3)
+            An array of positions in space, where the second dimension
+            corresponds to the three dimensions of the grid.
+
+        """
+
+        if hasattr(pos, "unit"):
+            pos = pos.si.value
+
+        # Find the bounds
+        if self.is_uniform:
+            ax0_min, ax0_max = np.min(self.ax0.si.value), np.max(self.ax0.si.value)
+            ax1_min, ax1_max = np.min(self.ax1.si.value), np.max(self.ax1.si.value)
+            ax2_min, ax2_max = np.min(self.ax2.si.value), np.max(self.ax2.si.value)
+
+        else:
+            pts0, pts1, pts2 = self.grids
+            ax0_min, ax0_max = np.min(self.pts0).si.value, np.max(self.pts0).si.value
+            ax1_min, ax1_max = np.min(self.pts1).si.value, np.max(self.pts1).si.value
+            ax2_min, ax2_max = np.min(self.pts2).si.value, np.max(self.pts2).si.value
+
+        # Check each point elementwise against the bounds
+        on_grid = (
+            np.greater(ax0_min, pos[:, 0]).astype(np.int8)
+            + np.less(ax0_max, pos[:, 0]).astype(np.int8)
+            + np.greater(ax1_min, pos[:, 1]).astype(np.int8)
+            + np.less(ax1_max, pos[:, 1]).astype(np.int8)
+            + np.greater(ax2_min, pos[:, 2]).astype(np.int8)
+            + np.less(ax2_max, pos[:, 2]).astype(np.int8)
+        )
+
+        return np.where(on_grid == 0, True, False)
+
+    def vector_intersects(self, p1, p2):
+        r"""
+        Returns True if the vector from p1 to p2 intersects the grid. Otherwise,
+        returns false. This is a standard ray-box intersection algorithm.
+        """
+        p1, p2 = p1.si.value, p2.si.value
+        # Caclulate the minimum and maximum of each
+        Ax, Bx = np.min(self.pts0.si.value), np.max(self.pts0.si.value)
+        Ay, By = np.min(self.pts1.si.value), np.max(self.pts1.si.value)
+        Az, Bz = np.min(self.pts2.si.value), np.max(self.pts2.si.value)
+        A = np.array([Ax, Ay, Az])
+        B = np.array([Bx, By, Bz])
+
+        # Calculate the equation of the line from p1 to p2 such that
+        # r = p1 + t*D
+        D = np.abs(p2 - p1)
+
+        # Calculate the intersection points. These operations are just vectorized
+        # for convenience. Ignore div-by-zero: outputting infty's here is fine.
+        with np.errstate(divide="ignore"):
+            Tmin = (A - p1) / D
+            Tmax = (B - p1) / D
+
+        Tmin = np.max(Tmin)
+        Tmax = np.min(Tmax)
+
+        return Tmin < Tmax
+
     # *************************************************************************
     # Interpolators
     # *************************************************************************
 
+    # These arrays are used to enable persistent interpolation (faster repeated
+    # calls with the same grid and arguments)
+
+    _interp_args = []
+    _interp_quantities = None
+    _interp_units = None
```

Comment:
I wonder if we could use https://docs.python.org/3/library/functools.html#functools.cached_property to simplify this part? I've been using it recently and it simplifies such code by a ton. It is 3.8 only, but there is a backport https://pypi.org/project/cached-property/ that I would be more than happy to add to our requirements until we're 3.8 only!

Reply:
I will look into this!! I can always count on you knowing a more efficient way of doing things :)